### PR TITLE
Improve the logic of kiting enemies with archers

### LINKED ISSUE #7607: The AI tries to kite/move away its archers from creatures it can't escape from
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

The battle AI sometimes makes the decision to kite (try to outrun and gain a free shot) when it clearly will not work.
This happens against both flyers and other creatures that will be able to get to them in 1 turn.

Here's an example of 100 war trolls against 350 gargoyles. I removed basic leadership, even if trying to kite to gain a bonus turn is a very questionable decision.

Flyer:

https://github.com/ihhub/fheroes2/assets/12501091/33c594ec-e45d-4faf-a395-79d6e1f24138

Walker 150 master druids vs 500 boars:

https://github.com/ihhub/fheroes2/assets/12501091/bbf3b7da-41d6-488c-aedf-474205ef018e

Here is the battle log from the above battle with the gargoyles. I've marked the places where the kiting decision comes in with an arrow ----> :

<details>

```
Battle::Loader:  army1 color(Blue), commander(Lord Kilburn): 100 War Trolls,
22.08.2023 17:01:41: [DBG_BATTLE]       Battle::Loader:  army2 color(None), : 350 Gargoyles,
22.08.2023 17:01:41: [DBG_BATTLE]       Battle::Arena::Turns:  1
22.08.2023 17:01:41: [DBG_BATTLE]       Battle::Arena::TurnTroop:  Unit: [ 350 Gargoyles, None, pos: 32, -1, reflect, mode(0x00000000), uid(0x00000002), speed(Very Fast, 6), hp(5250), died(0) ]
22.08.2023 17:01:41: [DBG_BATTLE]       AI::BattlePlanner::analyzeBattleState:  Tactic 0 chosen. Archers: 0, vs enemy 6614.29 ratio is 1
22.08.2023 17:01:41: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles start their turn. Side: 0
22.08.2023 17:01:41: [DBG_BATTLE]       AI::BattlePlanner::meleeUnitOffense:  Gargoyles attacking War Trolls at 55
22.08.2023 17:01:41: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles melee phase end, target cell is 55
22.08.2023 17:01:41: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles moving to cell 55
22.08.2023 17:01:41: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles melee offense, focus enemy War Trolls threat level: 120000
22.08.2023 17:01:41: [DBG_BATTLE]       Battle::Arena::ApplyActionMove:  Unit: [ 350 Gargoyles, None, pos: 32, -1, reflect ], dst: 55, (head: 55, tail: -1)
22.08.2023 17:01:43: [DBG_BATTLE]       Battle::Arena::ApplyActionAttack:  Unit: [ 350 Gargoyles, None, pos: 55, -1, reflect ] to Unit: [ 100 War Trolls, Blue, pos: 44, -1 ]
22.08.2023 17:01:43: [DBG_BATTLE]       Battle::Arena::TargetsApplyDamage:  number of targets: 1
22.08.2023 17:01:43: [DBG_BATTLE]       Battle::Unit::ApplyDamage:  607 to Unit: [ 100 War Trolls, Blue, pos: 44, -1, reflect ] and killed: 15
22.08.2023 17:01:44: [DBG_BATTLE]       Battle::Arena::TargetsApplyDamage:  number of targets: 1
22.08.2023 17:01:44: [DBG_BATTLE]       Battle::Unit::ApplyDamage:  541 to Unit: [ 350 Gargoyles, None, pos: 55, -1 ] and killed: 36
22.08.2023 17:01:44: [DBG_BATTLE]       Battle::Arena::ApplyActionEnd:  Unit: [ 314 Gargoyles, None, pos: 55, -1, reflect ]
22.08.2023 17:01:44: [DBG_BATTLE]       Battle::Arena::TurnTroop:  Unit: [ 85 War Trolls, Blue, pos: 44, -1, mode(0x00000001), uid(0x00000001), speed(Fast, 5), hp(3393), died(15) ]
22.08.2023 17:01:45: [DBG_BATTLE]       Battle::Arena::ApplyActionAutoSwitch:  color: Blue, status: on
22.08.2023 17:01:45: [DBG_BATTLE]       AI::BattlePlanner::analyzeBattleState:  Tactic 1 chosen. Archers: 5622.15, vs enemy 0 ratio is 0
22.08.2023 17:01:45: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  War Trolls start their turn. Side: 1
----> 22.08.2023 17:01:45: [DBG_BATTLE]       AI::BattlePlanner::archerDecision:  War Trolls archer kiting enemy, target cell is 3
22.08.2023 17:01:45: [DBG_BATTLE]       AI::BattlePlanner::archerDecision:  War Trolls moving to cell 3
22.08.2023 17:01:45: [DBG_BATTLE]       Battle::Arena::ApplyActionMove:  Unit: [ 85 War Trolls, Blue, pos: 44, -1 ], dst: 3, (head: 3, tail: -1)
22.08.2023 17:01:45: [DBG_BATTLE]       Battle::Arena::GetPath:  34, 23, 13, 2, 3,
22.08.2023 17:01:46: [DBG_BATTLE]       Battle::Arena::ApplyActionEnd:  Unit: [ 85 War Trolls, Blue, pos: 3, -1 ]
22.08.2023 17:01:46: [DBG_BATTLE]       Battle::Arena::Turns:  2
22.08.2023 17:01:46: [DBG_BATTLE]       Battle::Arena::TurnTroop:  Unit: [ 314 Gargoyles, None, pos: 55, -1, reflect, mode(0x00000000), uid(0x00000002), speed(Very Fast, 6), hp(4709), died(36) ]
22.08.2023 17:01:46: [DBG_BATTLE]       AI::BattlePlanner::analyzeBattleState:  Tactic 0 chosen. Archers: 0, vs enemy 5622.15 ratio is 1
22.08.2023 17:01:46: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles start their turn. Side: 0
22.08.2023 17:01:46: [DBG_BATTLE]       AI::BattlePlanner::meleeUnitOffense:  Gargoyles attacking War Trolls at 15
22.08.2023 17:01:46: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles melee phase end, target cell is 15
22.08.2023 17:01:46: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles moving to cell 15
22.08.2023 17:01:46: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles melee offense, focus enemy War Trolls threat level: 102000
22.08.2023 17:01:46: [DBG_BATTLE]       Battle::Arena::ApplyActionMove:  Unit: [ 314 Gargoyles, None, pos: 55, -1, reflect ], dst: 15, (head: 15, tail: -1)
22.08.2023 17:01:47: [DBG_BATTLE]       Battle::Arena::ApplyActionAttack:  Unit: [ 314 Gargoyles, None, pos: 15, -1, reflect ] to Unit: [ 85 War Trolls, Blue, pos: 3, -1 ]
22.08.2023 17:01:48: [DBG_BATTLE]       Battle::Arena::TargetsApplyDamage:  number of targets: 1
22.08.2023 17:01:48: [DBG_BATTLE]       Battle::Unit::ApplyDamage:  763 to Unit: [ 85 War Trolls, Blue, pos: 3, -1 ] and killed: 19
22.08.2023 17:01:48: [DBG_BATTLE]       Battle::Arena::TargetsApplyDamage:  number of targets: 1
22.08.2023 17:01:48: [DBG_BATTLE]       Battle::Unit::ApplyDamage:  395 to Unit: [ 314 Gargoyles, None, pos: 15, -1, reflect ] and killed: 26
22.08.2023 17:01:49: [DBG_BATTLE]       Battle::Arena::ApplyActionEnd:  Unit: [ 288 Gargoyles, None, pos: 15, -1, reflect ]
22.08.2023 17:01:49: [DBG_BATTLE]       Battle::Arena::TurnTroop:  Unit: [ 66 War Trolls, Blue, pos: 3, -1, mode(0x00000001), uid(0x00000001), speed(Fast, 5), hp(2637), died(34) ]
22.08.2023 17:01:49: [DBG_BATTLE]       AI::BattlePlanner::analyzeBattleState:  Tactic 1 chosen. Archers: 4365.43, vs enemy 0 ratio is 0
22.08.2023 17:01:49: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  War Trolls start their turn. Side: 1
----->  22.08.2023 17:01:49: [DBG_BATTLE]       AI::BattlePlanner::archerDecision:  War Trolls archer kiting enemy, target cell is 8
22.08.2023 17:01:49: [DBG_BATTLE]       AI::BattlePlanner::archerDecision:  War Trolls moving to cell 8
22.08.2023 17:01:49: [DBG_BATTLE]       Battle::Arena::ApplyActionMove:  Unit: [ 66 War Trolls, Blue, pos: 3, -1 ], dst: 8, (head: 8, tail: -1)
22.08.2023 17:01:49: [DBG_BATTLE]       Battle::Arena::GetPath:  4, 5, 6, 7, 8,
22.08.2023 17:01:50: [DBG_BATTLE]       Battle::Arena::ApplyActionEnd:  Unit: [ 66 War Trolls, Blue, pos: 8, -1 ]
22.08.2023 17:01:50: [DBG_BATTLE]       Battle::Arena::Turns:  3
22.08.2023 17:01:50: [DBG_BATTLE]       Battle::Arena::TurnTroop:  Unit: [ 288 Gargoyles, None, pos: 15, -1, reflect, mode(0x00000000), uid(0x00000002), speed(Very Fast, 6), hp(4314), died(62) ]
22.08.2023 17:01:50: [DBG_BATTLE]       AI::BattlePlanner::analyzeBattleState:  Tactic 0 chosen. Archers: 0, vs enemy 4365.43 ratio is 1
22.08.2023 17:01:50: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles start their turn. Side: 0
22.08.2023 17:01:50: [DBG_BATTLE]       AI::BattlePlanner::meleeUnitOffense:  Gargoyles attacking War Trolls at 7
22.08.2023 17:01:50: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles melee phase end, target cell is 7
22.08.2023 17:01:50: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles moving to cell 7
22.08.2023 17:01:50: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles melee offense, focus enemy War Trolls threat level: 79200
22.08.2023 17:01:50: [DBG_BATTLE]       Battle::Arena::ApplyActionMove:  Unit: [ 288 Gargoyles, None, pos: 15, -1, reflect ], dst: 7, (head: 7, tail: -1)
22.08.2023 17:01:51: [DBG_BATTLE]       Battle::Arena::ApplyActionAttack:  Unit: [ 288 Gargoyles, None, pos: 7, -1, reflect ] to Unit: [ 66 War Trolls, Blue, pos: 8, -1 ]
22.08.2023 17:01:51: [DBG_BATTLE]       Battle::Arena::TargetsApplyDamage:  number of targets: 1
22.08.2023 17:01:51: [DBG_BATTLE]       Battle::Unit::ApplyDamage:  663 to Unit: [ 66 War Trolls, Blue, pos: 8, -1, reflect ] and killed: 16
22.08.2023 17:01:52: [DBG_BATTLE]       Battle::Arena::TargetsApplyDamage:  number of targets: 1
22.08.2023 17:01:52: [DBG_BATTLE]       Battle::Unit::ApplyDamage:  274 to Unit: [ 288 Gargoyles, None, pos: 7, -1 ] and killed: 18
22.08.2023 17:01:52: [DBG_BATTLE]       Battle::Arena::ApplyActionEnd:  Unit: [ 270 Gargoyles, None, pos: 7, -1, reflect ]
22.08.2023 17:01:52: [DBG_BATTLE]       Battle::Arena::TurnTroop:  Unit: [ 50 War Trolls, Blue, pos: 8, -1, mode(0x00000001), uid(0x00000001), speed(Fast, 5), hp(1977), died(50) ]
22.08.2023 17:01:52: [DBG_BATTLE]       AI::BattlePlanner::analyzeBattleState:  Tactic 1 chosen. Archers: 3307.14, vs enemy 0 ratio is 0
22.08.2023 17:01:52: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  War Trolls start their turn. Side: 1
22.08.2023 17:01:52: [DBG_BATTLE]       AI::BattlePlanner::archerDecision:  War Trolls archer decided to fight back: -221
22.08.2023 17:01:52: [DBG_BATTLE]       Battle::Arena::ApplyActionAttack:  Unit: [ 50 War Trolls, Blue, pos: 8, -1 ] to Unit: [ 270 Gargoyles, None, pos: 7, -1, reflect ]
22.08.2023 17:01:52: [DBG_BATTLE]       Battle::Arena::TargetsApplyDamage:  number of targets: 1
22.08.2023 17:01:52: [DBG_BATTLE]       Battle::Unit::ApplyDamage:  319 to Unit: [ 270 Gargoyles, None, pos: 7, -1 ] and killed: 21
22.08.2023 17:01:53: [DBG_BATTLE]       Battle::Arena::TargetsApplyDamage:  number of targets: 1
22.08.2023 17:01:53: [DBG_BATTLE]       Battle::Unit::ApplyDamage:  582 to Unit: [ 50 War Trolls, Blue, pos: 8, -1, reflect ] and killed: 15
22.08.2023 17:01:53: [DBG_BATTLE]       Battle::Arena::ApplyActionEnd:  Unit: [ 35 War Trolls, Blue, pos: 8, -1 ]
22.08.2023 17:01:53: [DBG_BATTLE]       Battle::Arena::Turns:  4
22.08.2023 17:01:53: [DBG_BATTLE]       Battle::Arena::TurnTroop:  Unit: [ 249 Gargoyles, None, pos: 7, -1, reflect, mode(0x00000000), uid(0x00000002), speed(Very Fast, 6), hp(3721), died(101) ]
22.08.2023 17:01:53: [DBG_BATTLE]       AI::BattlePlanner::analyzeBattleState:  Tactic 0 chosen. Archers: 0, vs enemy 2315 ratio is 1
22.08.2023 17:01:53: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles start their turn. Side: 0
22.08.2023 17:01:53: [DBG_BATTLE]       AI::BattlePlanner::meleeUnitOffense:  Gargoyles attacking War Trolls at 19
22.08.2023 17:01:53: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles melee phase end, target cell is 19
22.08.2023 17:01:53: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles moving to cell 19
22.08.2023 17:01:53: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles melee offense, focus enemy War Trolls threat level: 21000
22.08.2023 17:01:53: [DBG_BATTLE]       Battle::Arena::ApplyActionMove:  Unit: [ 249 Gargoyles, None, pos: 7, -1, reflect ], dst: 19, (head: 19, tail: -1)
22.08.2023 17:01:53: [DBG_BATTLE]       Battle::Arena::ApplyActionAttack:  Unit: [ 249 Gargoyles, None, pos: 19, -1, reflect ] to Unit: [ 35 War Trolls, Blue, pos: 8, -1 ]
22.08.2023 17:01:54: [DBG_BATTLE]       Battle::Arena::TargetsApplyDamage:  number of targets: 1
22.08.2023 17:01:54: [DBG_BATTLE]       Battle::Unit::ApplyDamage:  501 to Unit: [ 35 War Trolls, Blue, pos: 8, -1, reflect ] and killed: 12
22.08.2023 17:01:54: [DBG_BATTLE]       Battle::Arena::TargetsApplyDamage:  number of targets: 1
22.08.2023 17:01:54: [DBG_BATTLE]       Battle::Unit::ApplyDamage:  124 to Unit: [ 249 Gargoyles, None, pos: 19, -1 ] and killed: 9
22.08.2023 17:01:55: [DBG_BATTLE]       Battle::Arena::ApplyActionEnd:  Unit: [ 240 Gargoyles, None, pos: 19, -1, reflect ]
22.08.2023 17:01:55: [DBG_BATTLE]       Battle::Arena::TurnTroop:  Unit: [ 23 War Trolls, Blue, pos: 8, -1, mode(0x00000001), uid(0x00000001), speed(Fast, 5), hp(899), died(77) ]
22.08.2023 17:01:55: [DBG_BATTLE]       AI::BattlePlanner::analyzeBattleState:  Tactic 1 chosen. Archers: 1521.29, vs enemy 0 ratio is 0
22.08.2023 17:01:55: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  War Trolls start their turn. Side: 1
22.08.2023 17:01:55: [DBG_BATTLE]       AI::BattlePlanner::archerDecision:  War Trolls archer decided to fight back: -314
22.08.2023 17:01:55: [DBG_BATTLE]       Battle::Arena::ApplyActionAttack:  Unit: [ 23 War Trolls, Blue, pos: 8, -1 ] to Unit: [ 240 Gargoyles, None, pos: 19, -1, reflect ]
22.08.2023 17:01:55: [DBG_BATTLE]       Battle::Arena::TargetsApplyDamage:  number of targets: 1
22.08.2023 17:01:55: [DBG_BATTLE]       Battle::Unit::ApplyDamage:  145 to Unit: [ 240 Gargoyles, None, pos: 19, -1 ] and killed: 9
22.08.2023 17:01:56: [DBG_BATTLE]       Battle::Arena::TargetsApplyDamage:  number of targets: 1
22.08.2023 17:01:56: [DBG_BATTLE]       Battle::Unit::ApplyDamage:  531 to Unit: [ 23 War Trolls, Blue, pos: 8, -1, reflect ] and killed: 13
22.08.2023 17:01:56: [DBG_BATTLE]       Battle::Arena::ApplyActionEnd:  Unit: [ 10 War Trolls, Blue, pos: 8, -1 ]
22.08.2023 17:01:56: [DBG_BATTLE]       Battle::Arena::Turns:  5
22.08.2023 17:01:56: [DBG_BATTLE]       Battle::Arena::TurnTroop:  Unit: [ 231 Gargoyles, None, pos: 19, -1, reflect, mode(0x00000000), uid(0x00000002), speed(Very Fast, 6), hp(3452), died(119) ]
22.08.2023 17:01:56: [DBG_BATTLE]       AI::BattlePlanner::analyzeBattleState:  Tactic 0 chosen. Archers: 0, vs enemy 661.429 ratio is 1
22.08.2023 17:01:56: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles start their turn. Side: 0
22.08.2023 17:01:56: [DBG_BATTLE]       AI::BattlePlanner::meleeUnitOffense:  Gargoyles attacking War Trolls at 9
22.08.2023 17:01:56: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles melee phase end, target cell is 9
22.08.2023 17:01:56: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles moving to cell 9
22.08.2023 17:01:56: [DBG_BATTLE]       AI::BattlePlanner::planUnitTurn:  Gargoyles melee offense, focus enemy War Trolls threat level: 6000
22.08.2023 17:01:56: [DBG_BATTLE]       Battle::Arena::ApplyActionMove:  Unit: [ 231 Gargoyles, None, pos: 19, -1, reflect ], dst: 9, (head: 9, tail: -1)
22.08.2023 17:01:56: [DBG_BATTLE]       Battle::Arena::ApplyActionAttack:  Unit: [ 231 Gargoyles, None, pos: 9, -1, reflect ] to Unit: [ 10 War Trolls, Blue, pos: 8, -1 ]
22.08.2023 17:01:56: [DBG_BATTLE]       Battle::Arena::TargetsApplyDamage:  number of targets: 1
22.08.2023 17:01:56: [DBG_BATTLE]       Battle::Unit::ApplyDamage:  441 to Unit: [ 10 War Trolls, Blue, pos: 8, -1 ] and killed: 10
22.08.2023 17:01:57: [DBG_BATTLE]       Battle::Unit::PostKilledAction:  Unit: [ 0 War Trolls, Blue, pos: 8, -1 ]
22.08.2023 17:02:00: [DBG_BATTLE]       Battle::Loader:  army1 color(Blue), commander(Lord Kilburn):
22.08.2023 17:02:00: [DBG_BATTLE]       Battle::Loader:  army2 color(None), : 231 Gargoyles,
22.08.2023 17:02:00: [DBG_BATTLE]       Battle::Loader:  army1: loss, army2: wins
```
</details>

### Save file

Just set up the battle in the battle mode. I found the above setup to be best at reproducing this behavior. 

### Additional info

I suggest to never try to kite flyers, UNLESS they have slow on them in which case they will actually move shorter.
In general the battle AI just needs to know when a creature will be able to get to them and not.

Here is the relevant part of the code:

https://github.com/ihhub/fheroes2/blob/0c08d956c54c299b9065092c757288cd33fc7cfe/src/fheroes2/ai/normal/ai_normal_battle.cpp#L604-L619

### LINKED ISSUE #7608: The AI should preemptively kite creatures to avoid getting hit
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

Currently the battle AI will wait until it has been struck to decide to kite, instead of kiting beforehand to avoid getting hit.

I found it reproducible with 150 greater druids against 350 golems:

https://github.com/ihhub/fheroes2/assets/12501091/d2a55cfb-bc0d-40d0-a98f-6f81b26b14ff

### Describe the possible solution

Make the AI kite before getting struck if it is within the range of the opposing creature.

### Additional info

Here is the relevant part of the code:

https://github.com/ihhub/fheroes2/blob/0c08d956c54c299b9065092c757288cd33fc7cfe/src/fheroes2/ai/normal/ai_normal_battle.cpp#L604-L619

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
close #7607
close #7608

https://github.com/ihhub/fheroes2/assets/32623900/104eed96-faa6-4f6a-96a2-0b39d902c5e6

Currently the logic is as follows: if stack of archers is in danger (i.e. it can be attacked during the current or next turn by one or more enemy stacks) but there are safe positions on the battlefield (i.e. positions that are not reachable for enemy stacks during one turn), then this stack will retreat to one of these safe positions, which is located at the maximum distance from all enemy stacks, unless there is only one enemy stack left that is guaranteed to be killed by this stack of archers. This may not always work 100% reliably, because both enemy and friendly stacks can move across the battlefield, and as a result, shorter paths to the same positions may appear.

Maybe it doesn't work perfectly in all possible cases, but, IMHO, it's better than what it is now. Perhaps there will be some specific ideas on how to make kiting even better, but so far there are none.

---

Also this PR implements the ability of archers with `AREA_SHOT` capability to attack tail cells of wide units if necessary. With this PR:

https://github.com/ihhub/fheroes2/assets/32623900/1b645482-b733-434c-abde-1d437a32b9b9

`master` branch:

https://github.com/ihhub/fheroes2/assets/32623900/af425f07-138d-4bcc-acc0-51989c6cbdaa

Note that with the second attack, Lich cannot hit all three targets.

### CODE REVIEW

**Comment by oleg-derevenetz** [comment]
@zenseii

> Also, not an easily reproducible one, but from your video in the opening statement I wonder why the archers choose to flee in this case and not kill the stack of 2 zombies, taking consideration your description in the opening statement, about when an archer will decide to just attack and not run, if it can take out a stack in one attack with 100% certainty.

Because for now they will choose to fight only if there is just a single enemy stack left on the battlefield, and they can kill this stack with 100% probability (well, not certainly 100% because of possible bad luck event). Because while they are shooting at one stack to kill it, the other stack(s) may get so close that the archers will no longer be able to retreat, and it is quite difficult to reliably calculate, because this is no longer a forecast just for the next turn. Therefore, for reliability, they will retreat in this case.

**Comment by oleg-derevenetz** [comment]
@zenseii

> The logic to not flee could be something dependant on whether the archer would be able to kill the attacking unit by retaliating? I can easily see how this could be problematic though and likely more complex, like taking into consideration the HP loss of being attacked (1hp lost from 1 peasant is obviously not a big deal for 300++ titans).

There will always be some corner cases like this. There is another problem as well - what if there are more than just one weak stacks on the battlefield, and although the Titan can still retreat at the moment and then shoot, but on the next turn, when he is blocked by the remaining weak stacks, he will no longer be able to retreat or even shoot? Again, it is quite difficult to reliably calculate this if we are talking about several turns ahead.

**Comment by oleg-derevenetz** [comment]
@zenseii

> Take into consideration this video where I show how you can easily manipulate a much stronger archer to not make any attack against just 5 skeletons, by carefully moving the skeletons so that you trick the titan into thinking it can escape.

Yes, but what are the formal parameters when to use this tactic of retreat, and when not to use? In theory, we can, for example, use this tactic if the enemy army is obviously stronger (how much?), and otherwise not use it.

**Comment by oleg-derevenetz** [comment]
@zenseii OK, I tried a slightly different logic, I suggest trying to play with it a little in various situations.

**Comment by oleg-derevenetz** [comment]
Hi @zenseii

> One big issue I found is that creatures are allowed to do a melee attack at a creature that is not adjacent to them.

Yeah, good catch, thank you! Should be fixed now.

**Comment by oleg-derevenetz** [comment]
@zenseii

> I believe we have an issue tracked for that. I'll try to look for it and link it. I'll test your latest changes too.

There should be no active issues of this kind regarding to battles. Any battle should always give the same results during the instant battle and during the manual battle if we press the "AUTO" button immediately on the first turn. And I want it to stay that way :)

**Comment by oleg-derevenetz** [comment]
Hi @ZENSEII, thank you very much! 

> As long as there are no cases when there is an infinite battle loop then it's fine, and I haven't been able to produce that.

There is a built-in protection from unlimited battles in the AI code, because they are possible even without this PR (for example, the famous "troll vs troll" case): if there were 50 turns in a row without deaths, then, if attacker is controlled by AI, he will retreat from battle, and if he is controlled by a human player, but the auto mode is used, then the auto mode will be forcibly turned off for him.

**Review by Districh-ru — APPROVED** [review]
Well done, @oleg-derevenetz!
I tested the battles in Battle Only mode and let AI to fully play one map, watching the battles, and I like this logic!
I tested the battles in the Battle Only mode and let AI to play fully one map watching the the battles, and I like this logic!

If players will learn it, than in some (rare) cases they can use it against the archers. :)

https://github.com/ihhub/fheroes2/assets/113276641/c3b3e0d1-e032-4f72-b71b-57e1bb01d9d0

This is not a bug of this logic. We cannot predict all cases. And learning AI behavior patterns is also a part of the game. :)
But later we can try to improve the logic, for example by setting limits on the number of consecutive retreat moves.

**Comment by oleg-derevenetz** [comment]
> If players will learn it, than in some (rare) cases they can use it against the archers. :)

The fact that you can scare a single stack of shooters with your own strong stack is not so scary. The main thing is that you can't scare a strong stack of shooters with your weak stack at the beginning of the game in order to abuse it and really take away, for example, an artifact guarded by a titan using only L2 magic and skeletons.

**Review by ihhub — COMMENTED** [review]
Hi @oleg-derevenetz , I left one tiny suggestion for the future. The rest looks good for me.

File: src/fheroes2/ai/normal/ai_normal_battle.cpp [hunk 1]
```diff
@@ -566,101 +534,379 @@ namespace AI
...
+                    }
+
+                    // The event of bad luck is deliberately not taken into account here, so that it does not look like
+                    // cheating, because a human player in a similar situation does not know about this event in advance
+                    const uint32_t guaranteedDamage = currentUnit.Modes( SP_BLESS ) ? currentUnit.CalculateMaxDamage( *enemy ) : currentUnit.CalculateMinDamage( *enemy );
+                    const uint32_t guaranteedKills = enemy->HowManyWillBeKilled( guaranteedDamage );
+
+                    assert( guaranteedKills <= enemy->GetCount() );
+
+                    if ( guaranteedKills != enemy->GetCount() ) {
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply src/fheroes2/ai/normal/ai_normal_battle.cpp:708]: Hi @ihhub this check is for that very case when human player tries to scare stack of mighty shooters by fast but weak stacks to force the shooters to move OR for a case when the last battered stack is left which shooters can just kill right there right now and no need to retreat from it. There is no need for another TODO here IMHO, if there will be some troubles and ideas how to fix these troubles, then we will change something.

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
@oleg-derevenetz thank you so much for these great changes!

**Pull request merged**